pkg/sql/sem/tree/eval.go

```diff
@@ -3156,17 +3149,33 @@ func PerformCast(ctx *EvalContext, d Datum, t coltypes.CastTargetType) (Datum, e
 		}
 
 	case *coltypes.TInterval:
-		// TODO(knz): Interval from float, decimal.
 		switch v := d.(type) {
 		case *DString:
 			return ParseDInterval(string(*v))
 		case *DCollatedString:
 			return ParseDInterval(v.Contents)
 		case *DInt:
-			// An integer duration represents a duration in microseconds.
-			return &DInterval{Duration: duration.Duration{Nanos: int64(*v) * 1000}}, nil
+			return &DInterval{Duration: duration.FromInt64(int64(*v))}, nil
+		case *DFloat:
+			return &DInterval{Duration: duration.FromFloat64(float64(*v))}, nil
 		case *DTime:
 			return &DInterval{Duration: duration.Duration{Nanos: int64(*v) * 1000}}, nil
+		case *DDecimal:
+			d := ctx.getTmpDec()
+			dnanos := v.Decimal
+			dnanos.Exponent += 9
+			_, err := DecimalCtx.RoundToIntegralValue(d, &dnanos)
```

Comment:
This happens to guarantee that `d.Exponent = 0` because RoundToIntegralValue is implemented under the covers with `Quantize(d, x, 0)`. However I think it is better to call that explicitly in case RoundToIntegralValue ever changes that could allow `d.Exponent > 0`, which would make the below computation using d.Coeff incorrect since it would have some implied 0s at the end. Since this shouldn't ever round, we can also use ExactCtx. So I think it should be `ExactCtx.Quantize(d, &dnanos, 0)`

Reply:
I have changed to Quantize but I had to define a new context: `ExactCtx` has `Precision = 0` so `ExactCtx.Quantize` can never work unless the argument is itself zero.
Can you check I have done this properly?

Also `Quantize()` has a bunch of check after the call to `quantize()` which I believe are not needed here. Is there an alternate code path that would allow me to skip these checks? (Does this question even make sense?)